dex/farm/src/lib.rs

```diff
@@ -613,4 +613,207 @@ pub trait Farm:
+            context.decrease_reward_reserve();
+            self.calculate_initial_farming_amount(&mut context);
+            self.increase_reward_with_compounded_rewards(&mut context);
+            self.commit_changes(&context);
+
+            // Update the farm supply even though we dont burn the tokens by ourselves.
+            // The position will be burned and this supply has to be updated.
+            self.farm_token_supply()
+                .update(|x| *x -= &context.get_tx_input().get_payments().get_first().amount);
+        }
+
+        let mut payments = ManagedVec::<Self::Api, EsdtTokenPayment<Self::Api>>::new();
+        payments.push(context.get_tx_input().get_payments().get_first().clone());
+        payments.push(context.get_final_reward().unwrap().clone());
+
+        self.self_proxy(new_farm_address)
+            .migrate_from_old_farm(orig_caller_opt)
+            .with_multi_token_transfer(payments)
+            .execute_on_dest_context_custom_range(|_, after| (after - 1, after))
+    }
+
+    //This function is neccessary in new farms.
+    #[payable("*")]
+    #[endpoint(migrateFromOldFarm)]
+    fn migrate_from_old_farm(
+        &self,
+        #[var_args] orig_caller_opt: OptionalArg<ManagedAddress>,
+    ) -> EsdtTokenPayment<Self::Api> {
+        assert!(self, !self.farm_configuration().is_empty(), b"empty config");
+        let config = self.farm_configuration().get();
+        assert!(self, !config.is_old, b"bad config");
+
+        let caller = self.blockchain().get_caller();
+        assert!(self, caller == config.old_farm_address, b"bad caller");
+
+        let payments = self.call_value().all_esdt_transfers();
+        assert!(self, payments.len() == 2, b"bad payments len");
+
+        let old_position = payments.get(0).unwrap();
+        assert!(self, old_position.amount != 0u64, b"bad farm amount");
+
+        assert!(
+            self,
+            old_position.token_identifier == config.old_farm_token_id,
+            b"bad farm token id"
+        );
+
+        let reward = payments.get(1).unwrap();
+        assert!(self, reward.amount != 0u64, b"bad reward amount");
+
+        let reward_token_id = self.reward_token_id().get();
+        assert!(
+            self,
+            reward.token_identifier == reward_token_id,
+            b"bad reward token id"
+        );
+
+        // The actual work starts here
+        self.reward_reserve().update(|x| *x += &reward.amount);
+
+        let old_attrs: FarmTokenAttributes<Self::Api> = self
+            .blockchain()
+            .get_esdt_token_data(
+                &self.blockchain().get_sc_address(),
+                &old_position.token_identifier,
+                old_position.token_nonce,
+            )
+            .decode_attributes()
+            .unwrap();
+
+        // Do not call burn_farm_tokens since this farm tokens belong to other contract
+        // which already updated its farm token supply counter.
+        self.send().esdt_local_burn(
```

Comment:
we have to give burn roles to this contract. which is not implemented in the old farm.